examples/cevae.py

```diff
@@ -0,0 +1,311 @@
+    def __init__(self, args):
+        self.latent_dim = args.latent_dim
+        super().__init__()
+        self.x_nn = DiagNormalNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers +
+                                  [args.feature_dim])
+        self.y0_nn = BernoulliNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers)
+        self.y1_nn = BernoulliNet([args.latent_dim] +
+                                  [args.hidden_dim] * args.num_layers)
+        self.t_nn = BernoulliNet([args.latent_dim])
+
+    def forward(self, x, t=None, y=None, size=None):
+        if size is None:
+            size = x.size(0)
+        with pyro.plate("data", size, subsample=x):
+            z = pyro.sample("z", self.z_dist())
+            x = pyro.sample("x", self.x_dist(z), obs=x)
+            t = pyro.sample("t", self.t_dist(z), obs=t)
+            y = pyro.sample("y", self.y_dist(t, z), obs=y)
+        return y
+
+    def z_dist(self):
+        return dist.Normal(0, 1).expand([self.latent_dim]).to_event(1)
+
+    def x_dist(self, z):
+        loc, scale = self.x_nn(z)
+        return dist.Normal(loc, scale).to_event(1)
+
+    def y_dist(self, t, z):
+        # Parameters are not shared among t values.
+        logits0 = self.y0_nn(z)
+        logits1 = self.y1_nn(z)
+        logits = torch.where(t.bool(), logits1, logits0)
+        return dist.Bernoulli(logits=logits)
+
+    def t_dist(self, z):
+        logits = self.t_nn(z)
+        return dist.Bernoulli(logits=logits)
+
+
+class Guide(PyroModule):
+    """
+    Inference model for causal effect estimation.
+
+        t ~ p(t|x)      # treatment
+        y ~ p(y|t,x)    # outcome
+        z ~ p(t|y,t,x)  # latent confounder, an embedding
+
+    Each of these distributions is defined by a neural network.  The ``y`` and
+    ``z`` distributions are defined by disjoint pairs of neural networks
+    defining ``p(-|t=0,...)`` and ``p(-|t=1,...)``; this allows highly
+    imbalanced treatment.
+    """
+    def __init__(self, args):
+        self.latent_dim = args.latent_dim
+        super().__init__()
+        self.t_nn = BernoulliNet([args.feature_dim])
+        self.y_nn = FullyConnected([args.feature_dim] +
+                                   [args.hidden_dim] * args.num_layers)
+        self.y0_nn = BernoulliNet([args.hidden_dim])
+        self.y1_nn = BernoulliNet([args.hidden_dim])
+        self.z0_nn = DiagNormalNet([1 + args.feature_dim] +
+                                   [args.hidden_dim] * args.num_layers +
+                                   [args.latent_dim])
+        self.z1_nn = DiagNormalNet([1 + args.feature_dim] +
+                                   [args.hidden_dim] * args.num_layers +
+                                   [args.latent_dim])
+
+    def forward(self, x, t=None, y=None, size=None):
+        if size is None:
+            size = x.size(0)
+        with pyro.plate("data", size, subsample=x):
```

Comment:
nit: use `subsample_size=x.size(0)` explicitly here for clarity since `x` isn't actually an index

Reply:
Hmm, this version avoids triggering a `_Subsample` site, whereas the version `subsample_size=len(x)` would trigger extra sampling. While I realize this is a bit of a hack since `x` is not an index tensor, it is a hack that we often use to avoid the extra computation.